algebra-core/src/curves/batch_arith.rs

```diff
@@ -0,0 +1,294 @@
+use crate::{biginteger::BigInteger, AffineCurve, Field, Vec};
+use core::ops::Neg;
+use num_traits::Zero;
+
+/// We use a batch size that is big enough to amortise the cost of the actual inversion
+/// close to zero while not straining the CPU cache by generating and fetching from
+/// large w-NAF tables and slices [G]
+pub const BATCH_SIZE: usize = 4096;
+/// 0 == Identity; 1 == Neg; 2 == GLV; 3 == GLV + Neg
+pub const ENDO_CODING_BITS: usize = 2;
+
+#[inline(always)]
+pub fn decode_endo_from_u32(index_code: u32) -> (usize, u8) {
+    (
+        index_code as usize >> ENDO_CODING_BITS,
+        index_code as u8 % 4,
+    )
+}
+
+pub trait BatchGroupArithmetic
+where
+    Self: Sized + Clone + Copy + Zero + Neg<Output = Self>,
+{
+    type BBaseField: Field;
+
+    /*
+    We use the w-NAF method, achieving point density of approximately 1/(w + 1)
+    and requiring storage of only 2^(w - 1).
+    Refer to e.g. Improved Techniques for Fast Exponentiation, Section 4
+    Bodo M¨oller 2002. https://www.bmoeller.de/pdf/fastexp-icisc2002.pdf
+    */
+
+    /// Computes [[p, 3 * p, ..., (2^w - 1) * p], ..., [q, 3* q,  ..., ]]
+    /// We need to manipulate the offsets when using the table
+    fn batch_wnaf_tables(bases: &[Self], w: usize) -> Vec<Self> {
+        let half_size = 1 << (w - 1);
+        let batch_size = bases.len();
+
+        let zero = Self::zero();
+        let mut tables = vec![zero; half_size * batch_size];
+
+        let mut a_2 = bases.to_vec();
+        let mut tmp = bases.to_vec();
+
+        let instr = (0..batch_size).map(|x| x as u32).collect::<Vec<_>>();
+        Self::batch_double_in_place(&mut a_2, &instr[..], None);
+
+        for i in 0..half_size {
+            if i != 0 {
+                let instr = (0..batch_size)
+                    .map(|x| (x as u32, x as u32))
+                    .collect::<Vec<_>>();
+                Self::batch_add_in_place(&mut tmp, &mut a_2.to_vec()[..], &instr[..]);
+            }
+
+            for (elem_id, &p) in tmp.iter().enumerate() {
+                tables[elem_id * half_size + i] = p.clone();
+            }
+        }
+        tables
+    }
+
+    /// Computes the vectorised version of the wnaf integer recoding
+    /// Optionally takes a slice of booleans which indicate whether that
+    /// scalar is negative. If so, it negates the recoding.
+    /// Mutates scalars in place
+    fn batch_wnaf_opcode_recoding<BigInt: BigInteger>(
+        scalars: &mut [BigInt],
+        w: usize,
+        negate: Option<&[bool]>,
+    ) -> Vec<Vec<Option<i16>>> {
+        assert!(w > 0);
+        let batch_size = scalars.len();
+        let window_size: i16 = 1 << (w + 1);
+        let half_window_size: i16 = 1 << w;
+
+        let mut op_code_vectorised =
+            Vec::<Vec<Option<i16>>>::with_capacity(scalars[0].as_ref().len() * 64);
+
+        let mut all_none = false;
+
+        match negate {
+            None => {
```

Comment:
Probably these two match branches can be unified, perhaps with something like the following:

```rust
negate.unwrap_or((), |b| assert_eq!(b.len(), scalars.len()); // precompute bounds check
let bools = (0..scalars.len()).map(|_| negate.map_or(false, |b| b[i]]);
```

When `negate` is `None`, I expect this to optimize out unused branches without issues.